### PR TITLE
Sort output for SWIFT cell structure

### PR DESCRIPTION
**Pull request by robjmcgibbon**
This PR sorts the output spatially so it can be quickly read in using the same cell metadata contained in SWIFT snapshots. A new group containing the cell metadata has been added to the output file. 

The output was already being sorted (by halo index), so this change shouldn't have any impact on runtime.

### CODE REVIEW

**Comment by MatthieuSchaller** [comment]
I think this makes sense, yes.

Best would be to test it using `swiftsimio` or, before editing that package, via the code snippet here: https://swift.strw.leidenuniv.nl/docs/Snapshots/index.html#quick-access-to-particles-via-hash-tables

I don't think more info is striclty required but if there is anything else you can think of that would help downstream, it can be added too.

**Comment by MatthieuSchaller** [comment]
swiftsimio will use the `Units` block in the snapshot and apply them for this too btw. Don't think it matters here but that could be important if SOAP ends up being run in kpc or Msun for instance.

**Comment by JBorrow** [comment]
Why don't we just store this sorted already as a quadtree..?

**Comment by JBorrow** [comment]
With a little bit of extra metadata, we can make this a fully-realised quadtree that already exists. It would be transparent to swiftsimio (as the top-level cells would remain the same) but we can immediately extract a quadtree with a tiny bit more metadata. @MatthieuSchaller, shouldn't this be possible already with the swift snapshots..?

**Comment by MatthieuSchaller** [comment]
Provided the top-level grid is a power of 2, no?

**Comment by JBorrow** [comment]
No, there is no restriction on the top level grid size for an oct-tree. Otherwise SWIFT wouldn't work... You would just need to sort the haloes inside each TLC according to their position in an agreed-upon format for a quadtree.

**Comment by robjmcgibbon** [comment]
I've been able to load the SOAP file with swiftsimio after hacking it to detect the groups corresponding to each halo type

**Comment by robjmcgibbon** [comment]
I can add a cellid array if needed, but I feel it's easy enough to calculate that it's not really worth saving

**Comment by MatthieuSchaller** [comment]
Good for me.

File: category_filter.py [hunk 1]
```diff
@@ -127,9 +128,10 @@ def get_compression_metadata(self, property_output_name: str) -> Dict:
             if prop[0] == base_output_name:
                 compression = prop[6]
         if compression is None:
-            return {"Lossy Compression Algorithm": "None", "Is Compressed": False}
+
+            return {"Lossy compression filter": np.bytes_("None".encode('utf-8')), "Is Compressed": False}
         else:
-            return {"Lossy Compression Algorithm": compression, "Is Compressed": False}
+            return {"Lossy compression filter": np.bytes_(compression.encode('utf-8')), "Is Compressed": False}
```
Comment (JBorrow) [inline category_filter.py:134]: Why do this..?

File: combine_chunks.py [hunk 2]
```diff
@@ -90,15 +102,83 @@ def combine_chunks(
...
+            for name, value in cellgrid.constants.items():
+                const.attrs[name] = [value]
+
+            # Write cell information
+            cells = outfile.create_group("Cells")
+            cells_metadata = cells.create_group("Meta-data")
+            cells_metadata.attrs['dimension'] = cellgrid.dimension
+            cells_metadata.attrs['nr_cells'] = cellgrid.nr_cells
+            cell_size_cMpc = cellgrid.cell_size.to(cellgrid.boxsize.units).value
+            cells_metadata.attrs['size'] = cell_size_cMpc
```
Comment (JBorrow) [inline combine_chunks.py:169]: Is this right? I thought we saved this in the units of the file.

File: category_filter.py (hunk 1 above)
Comment (robjmcgibbon) [reply category_filter.py:134]: I have to force SOAP to write the string as bytes or else https://github.com/SWIFTSIM/swiftsimio/blob/fda13e2a7a704a37c6a9e26bfb7fee23ffc4b33f/swiftsimio/reader.py#L959 complains

File: combine_chunks.py (hunk 2 above)
Comment (robjmcgibbon) [reply combine_chunks.py:169]: I guess it's not guaranteed to be Mpc, but cellgrid.boxsize has the correct units, so I think this should be ok after I change the variable name.

File: category_filter.py (hunk 1 above)
Comment (robjmcgibbon) [reply category_filter.py:134]: Cool, I'm happy to do it that way

File: combine_chunks.py (hunk 2 above)
Comment (jchelly) [reply combine_chunks.py:169]: The ref_metadata parameter passed into this function contains unit information for all halo properties. We should look up the units there. If I run this branch in pdb then I can see that cellgrid.boxsize.units = a*snap_length but the units of cofp in ref_metadata are a*swift_mpc, which is not guaranteed to be the same.

**Comment by jchelly** [comment]
This pull request adds a Units group in the SOAP output which specifies units of length, mass, time etc and there's a comment asking if it's correct. I don't think it can be! The units of output halo properties are defined on a property by property basis in property_table.py and they're not consistent. We have lengths in kpc, cMpc and pMpc, for example. So to determine the units of a quantity you need to look at the attributes of that specific quantity. There's no set of base units which you can combine with the mass, length, time etc exponents to get the right units.

**Comment by JBorrow** [comment]
Hmm, that's interesting. That breaks significantly from the SWIFT convention where only the exponents are stored. How is that information stored in the HDF5 groups? Is it the same, just with a different CGS conversion factor..?

**Comment by jchelly** [comment]
Yes, to generate correct unyt units from soap output you need to look at the cgs conversion factor.

**Comment by JBorrow** [comment]
Cgs is a bit of a nasty unit system here because of potential overflow problems when using float32s... Are you using unyt internally in SOAP? We could have you write out the units as a string and then deserialize that with unyt on the other end.

**Comment by robjmcgibbon** [comment]
Summary of a discussion Matthieu and I had this morning
- I'll go through all the properties in SOAP and update their units so that their a-scale exponents are properly defined. This will allow them all to be correctly converted from co-moving to physical within swiftsimio
- I'll add a flag to units attributes of each property to specify whether it has been stored as co-moving or physical. This won't require any changes to swiftsimio other than reading this new attribute when creating a cosmo array for the dataset. I'll add this flag to the property table in SOAP.
- Add a flag to the units attributes of each dataset to specify whether conversion from physical to comoving is allowed.  This would require changing the cosmo_array object within swiftsimio to carry this flag. If someone tried to convert a property (such as ages) which they shouldn't, then swiftsimio can throw a warning/error. I'll also add this flag to all the properties we have in SOAP

**Comment by robjmcgibbon** [comment]
Spreadsheet with what I've set for the different classes of properties https://docs.google.com/spreadsheets/d/1-0VzX7mcnrtRp21Ouqx17nQUnQUUn5VzblV8ctvIYMU/edit?usp=sharing

**Comment by robjmcgibbon** [comment]
I've added the new flags to the hdf5 attributes: they're called `Value stored as physical` and `Property can be converted to comoving`, but I'm open to other names. 

I'm not outputting any properties as comoving for which a transformation to physical is not valid. I can't think of why this would be needed, but it might be better to call the second flag something like `Property can be converted between physical and comoving` so we support this possibility.

**Comment by MatthieuSchaller** [comment]
I think we can always do `comoving`-->`physical`. Apart from the case where we store in physical already. So the flag should really only be about whether something is convertible to comoving or not.

**Pull request merged**